Gifski/Intents.swift

```diff
@@ -1,6 +1,245 @@
+		let (dimensionsWidth, dimensionsHeight) = dimensions
+		switch mode {
+		case .exact:
+			guard let bottomLeftY,
+				  let x
+			else {
+				return nil
+			}
+			let cropWidth = self.width > 1 ? self.width : 1
+			let cropHeight = self.height > 1 ? self.height : 1
+
+			let topLeftY = dimensionsHeight - bottomLeftY - cropHeight
+			let entityRect = CGRect(x: x, y: topLeftY, width: cropWidth, height: cropHeight)
+			let videoRect = CGRect(origin: .zero, size: .init(width: dimensionsWidth, height: dimensionsHeight))
+			let intersectionRect = videoRect.intersection(entityRect)
+
+			guard intersectionRect.width >= 1,
+				  intersectionRect.height >= 1 else {
+				throw CropOutOfBounds(enteredRect: CGRect(x: x, y: bottomLeftY, width: cropWidth, height: cropHeight), videoRect: videoRect)
+			}
+			return CropRect(
+				x: Double(intersectionRect.x) / Double(dimensionsWidth),
+				y: Double(intersectionRect.y) / Double(dimensionsHeight),
+				width: Double(intersectionRect.width) / Double(dimensionsWidth),
+				height: Double(intersectionRect.height) / Double(dimensionsHeight)
+			)
+		case .aspectRatio:
+			let aspectWidth = self.width > 1 ? self.width : 1
+			let aspectHeight = self.height > 1 ? self.height : 1
+			return CropRect.centeredFrom(aspectWidth: Double(aspectWidth), aspectHeight: Double(aspectHeight), forDimensions: .init(width: dimensionsWidth, height: dimensionsHeight))
+		}
+	}
+	static func from(id: String) throws -> Self {
+		guard id != Self.errorId else {
+			return Self()
+		}
+		guard let data = Data(base64Encoded: id) else {
+			throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: [], debugDescription: "Invalid ID format"))
+		}
+		return try JSONDecoder().decode(Self.self, from: data)
+	}
+}
+
+
+struct CropEntityQuery: EntityQuery {
+	func entities(for identifiers: [CropEntity.ID]) async throws -> [CropEntity] {
+		try identifiers.map {
+			try CropEntity.from(id: $0)
+		}
+	}
+	func suggestedEntities() async throws -> [CropEntity] {
+		PickerAspectRatio.presets.map {
+			var crop  = CropEntity()
+			crop.mode = .aspectRatio
+			crop.width = $0.width
+			crop.height = $0.height
+			return crop
+		}
+	}
+}
+
+enum CropIntentMode: String, AppEnum, CaseIterable, Codable, Hashable {
+	static var caseDisplayRepresentations: [Self: DisplayRepresentation] = [
+		.aspectRatio: DisplayRepresentation(title: "a fixed aspect ratio"),
+		.exact: DisplayRepresentation(title: "exact dimensions")
+	]
+	case aspectRatio
+	case exact
+	static var typeDisplayRepresentation = TypeDisplayRepresentation(name: "Crop Mode")
+}
+struct CropOutOfBounds: LocalizedError, CustomNSError {
+	let enteredRect: CGRect
+	let videoRect: CGRect
+	var errorDescription: String? {
+		"Crop rectangle is out of bounds! It's \(enteredRect.width)x\(enteredRect.height) at (\(enteredRect.x),\(enteredRect.y)), but it needs to fit inside \(videoRect.width)x\(videoRect.height) for this video."
```

Comment:
Convert the values to `Int` so they are displayed without `.0` since they are always whole numbers.